File: libs/csharp/common/src/Core/Nodes/Scope.cs
using Crosslight.Core.Nodes;
using Crosslight.Core.Utilities;
using System.Runtime.InteropServices;
using Crosslight.Core;
using static Crosslight.Core.ILanguage;

namespace Crosslight.src.Core.Nodes
{
    /// <summary>
    /// Declaration scope, mapping to e.g. namespaces.
    /// </summary>
    /// <remarks>
    /// Empty scope is assumed to be global. If a truly empty scope is desired, an empty node can be placed inside.
    /// </remarks>
    public struct Scope : INodePayload
    {
        private struct Imported
        {
            public nint Identifier;
        }

        /// <summary>
        /// Full identifier of the scope, can be separated with '.'. Can be empty or <see langword="null"/>.
        /// </summary>
        public string? Identifier;

        /// <summary>
        /// Create a new managed instance of the scope payload from the native representation.
        /// </summary>
        /// <param name="pointer">The native pointer to a scope payload.</param>
        public Scope(nint pointer)
        {
            Imported imported = Marshal.PtrToStructure<Imported>(pointer);

            Identifier = imported.Identifier == 0
                ? null
                : Marshal.PtrToStringUTF8(imported.Identifier);
        }

        /// <inheritdoc/>
        public readonly NodeType Type => NodeType.Scope;

        /// <inheritdoc/>
        public readonly nint ToPointer(AcquireDelegate? acquire)
        {
            acquire ??= Marshal.AllocCoTaskMem;
            nint pointer = acquire(Marshal.SizeOf<Imported>());

            if (pointer == 0)
            {
                return 0;
            }

            nint identifier = Identifier == null
                ? 0
                : Utf8String.ToPointer(Identifier, acquire);

            Imported imported = new()
            {
                Identifier = identifier,
            };

            Marshal.StructureToPtr(imported, pointer, false);

            return pointer;
        }

        public readonly object? AcceptVisitor(Node node, INodePayloadVisitor visitor)
        {
            return visitor.VisitScope(node, this);
        }

        public override string ToString()
        {
            return $"{{ {nameof(Scope)}-{Identifier} }}";
        }
    }
}

File: libs/cpp/lang/src/builders/scope.cpp
#include "lang/builders/scope.hpp"
#include "lang/builders/allocator.hpp"
#include "lang/builders/builder.hpp"
#include "lang/utils.hpp"
#include "core/node.h"
#include "core/nodes/node_type.h"

namespace b = cl::lang::builders;

b::builder
b::scope(const b::allocator& m, const char *identifier) {
    return m.scope(identifier);
}

b::builder
b::allocator::scope(const char* identifier) const {
    auto* payload = static_cast<struct cl_node_scope*>(acquire(sizeof(struct cl_node_scope)));

    if (payload == nullptr) {
        return { *this, nullptr, nullptr };
    }

    payload->identifier = utils::string_duplicate(identifier, acquire);

    return builder::from_payload(*this, payload, ::scope);
}
